rgw/admin/user.go

```diff
@@ -0,0 +1,148 @@
+	Suspended   int           `json:"suspended" url:"suspended"`
+	MaxBuckets  int           `json:"max_buckets" url:"max-buckets"`
+	Subusers    []interface{} `json:"subusers"`
+	Keys        []struct {
+		User      string `json:"user"`
+		AccessKey string `json:"access_key" url:"access-key"`
+		SecretKey string `json:"secret_key" url:"secret-key"`
+	} `json:"keys"`
+	SwiftKeys []interface{} `json:"swift_keys"`
+	Caps      []struct {
+		Type string `json:"type"`
+		Perm string `json:"perm"`
+	} `json:"caps"`
+	OpMask              string        `json:"op_mask"`
+	DefaultPlacement    string        `json:"default_placement"`
+	DefaultStorageClass string        `json:"default_storage_class"`
+	PlacementTags       []interface{} `json:"placement_tags"`
+	BucketQuota         struct {
+		Enabled    bool `json:"enabled"`
+		CheckOnRaw bool `json:"check_on_raw"`
+		MaxSize    int  `json:"max_size"`
+		MaxSizeKb  int  `json:"max_size_kb"`
+		MaxObjects int  `json:"max_objects"`
+	} `json:"bucket_quota"`
+	UserQuota struct {
+		Enabled    bool `json:"enabled"`
+		CheckOnRaw bool `json:"check_on_raw"`
+		MaxSize    int  `json:"max_size"`
+		MaxSizeKb  int  `json:"max_size_kb"`
+		MaxObjects int  `json:"max_objects"`
+	} `json:"user_quota"`
+	TempURLKeys []interface{} `json:"temp_url_keys"`
+	Type        string        `json:"type"`
+	MfaIds      []interface{} `json:"mfa_ids"`
+	KeyType     string        `url:"key-type"`
+	Tenant      string        `url:"tenant"`
+	GenerateKey bool          `url:"generate-key"`
+	PurgeData   int           `url:"purge-data"`
+}
+
+// GetUser retrieves a given object store user
+func (api *API) GetUser(ctx context.Context, user User) (*User, error) {
+	if user.UserID == "" {
+		return nil, errors.New("missing user UID")
+	}
+
+	body, err := api.callAPI(ctx, get, "/user", valueToURLParams(user))
+	if err != nil {
+		return nil, err
+	}
+
+	u := &User{}
+	err = json.Unmarshal(body, &u)
+	if err != nil {
+		return nil, fmt.Errorf("%s. %s. %w", unmarshalWrapError, string(body), err)
+	}
+
+	return u, nil
+}
+
+// GetUsers lists all object store users
+func (api *API) GetUsers(ctx context.Context) (*[]string, error) {
+	body, err := api.callAPI(ctx, get, "/metadata/user", nil)
+	if err != nil {
+		return nil, err
+	}
+	var users *[]string
+	err = json.Unmarshal(body, &users)
+	if err != nil {
+		return nil, fmt.Errorf("%s. %s. %w", unmarshalWrapError, string(body), err)
+	}
+
+	return users, nil
+}
+
+// CreateUser creates a user in the object store
+func (api *API) CreateUser(ctx context.Context, user User) (*User, error) {
+	if user.UserID == "" {
+		return nil, errors.New("missing user UID")
+	}
+	if user.DisplayName == "" {
+		return nil, errors.New("missing user display name")
+	}
+
+	// Send request
+	body, err := api.callAPI(ctx, put, "/user", valueToURLParams(user))
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal response into Go type
+	u := &User{}
+	err = json.Unmarshal(body, &u)
```

Comment:
When trying to run the tests locally I triggered an issue, which in turn caused the HTTP call to return a 404. However, I noticed that all the functions end up returning errors related to unmarshalling because even though the http call "failed" the function tried unmarshalling 404-error xml as json.

Again, this falls into the category of not-strictly-needed-for-this-pr, but I think it may make sense to check the http codes in some of these apis prior to trying to decode json (or perhaps checking the content type header would be even more "elegant"??)

Reply:
I believe some earlier versions of Ceph were not outputting json by default. I should fix that.

Reply:
@phlogistonjohn  Can you give a me reproducer of that issue? Thanks!

Reply:
I changed the code so that rgw was listening on localhost and connected to it via localhost but did not change the "rgw dns name" value in the config. This causes 404s complaining about the bucket not existing, and the body of these responses are in XML, not JSON.

Reply:
If that is not clear/specific enough, just say so and I'll send it as code/patch, but I think that should be simple enough to repro from that comment.

Reply:
Ohhh now I get it :-). I'll handle that case too, it's when error comes from rgw itself not the admin ops API.

Reply:
Hum I don't see to find a good repro with the integration tests, however with a simple curl I think I'm getting the same error as you did:

```
[root@5df921d1a642 go-ceph]# curl 127.0.0.1:80
<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>[root@5df921d1a642 go-ceph]# %
 ```

Did you get the error while using the library?

Reply:
Yeah, here's a more worked out reproducer:
Diff:
```diff
diff --git a/Makefile b/Makefile
index 5e9caea..ecd25b0 100644
--- a/Makefile
+++ b/Makefile
@@ -46,7 +46,7 @@ test:
 .PHONY: test-docker test-container test-multi-container
 test-docker: test-container
 test-container: $(BUILDFILE) $(RESULTS_DIR)
-       $(CONTAINER_CMD) run $(CONTAINER_OPTS) --rm -e EXCLUDED_PKGS="contrib|rgw/admin" -v $(CURDIR):/go/src/github.com/ceph/go-ceph$(VOLUME_FLAGS) $(RESULTS_VOLUME) $(CI_IMAGE_TAG) $(ENTRYPOINT_ARGS)
+       $(CONTAINER_CMD) run $(CONTAINER_OPTS) --rm -e EXCLUDED_PKGS="contrib" -v $(CURDIR):/go/src/github.com/ceph/go-ceph$(VOLUME_FLAGS) $(RESULTS_VOLUME) $(CI_IMAGE_TAG) $(ENTRYPOINT_ARGS)
 test-multi-container: $(BUILDFILE) $(RESULTS_DIR)
        $(CONTAINER_CMD) kill test_ceph_a test_ceph_b 2>/dev/null || true
        $(CONTAINER_CMD) volume remove test_ceph_a_data test_ceph_b_data 2>/dev/null || true
diff --git a/rgw/admin/radosgw_test.go b/rgw/admin/radosgw_test.go
index d4f8bd6..98270ee 100644
--- a/rgw/admin/radosgw_test.go
+++ b/rgw/admin/radosgw_test.go
@@ -21,7 +21,7 @@ func TestRadosGWTestSuite(t *testing.T) {
 func (suite *RadosGWTestSuite) SetupConnection() {
        suite.accessKey = "2262XNX11FZRR44XWIRD"
        suite.secretKey = "rmtuS1Uj1bIC08QFYGW18GfSHAbkPqdsuYynNudw"
-       suite.endpoint = "http://test_ceph_a"
+       suite.endpoint = "http://localhost"
 }
```

Command
```
make test-container  CEPH_VERSION=pacific ENTRYPOINT_ARGS="--test-pkg=rgw/admin  "  RESULTS_DIR="$PWD/_results"
```

I don't think CEPH_VERSION or RESULTS_DIR matter, but in case they do these are the values I used and got the same set of errors as before. Not every test case fails. A couple that do:  TestRadosGWTestSuite/TestUser, TestRadosGWTestSuite/TestUsage